test/integration/tls/tls_stack.go

```diff
@@ -0,0 +1,908 @@
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"net/http"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"strings"
+	"time"
+
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/types"
+)
+
+const (
+	//If you're not able to access below images, try to build them based on the app code under "timeout_app"
+	//directory and push it to any accessible ECR repo and update the below values
+	defaultFrontEndImage = "928111597794.dkr.ecr.us-west-2.amazonaws.com/amazon/tls-e2e:feapp-tlsimage"
+	defaultBackEndImage  = "928111597794.dkr.ecr.us-west-2.amazonaws.com/amazon/tls-e2e:beapp-tlsimage"
+
+	tlsTest          = "tls-e2e"
+	AppContainerPort = 8080
+
+	tlsConnectionError         = "upstream connect error"
+	expectedBackendResponse = "backend"
+)
+
+//FrontEnd -> Will have TLS Validation enabled in both the test cases
+//BackEnd  -> TLS support will be toggled between the test cases to verify that the communication is possible only
+//            when the TLS handshake works between FrontEnd and BackEnd envoys
+
+type TLSStack struct {
+	// service discovery type
+	ServiceDiscoveryType manifest.ServiceDiscoveryType
+
+	// ====== runtime variables ======
+	mesh      *appmesh.Mesh
+	namespace *corev1.Namespace
+
+	FrontEndVN *appmesh.VirtualNode
+	FrontEndDP *appsv1.Deployment
+
+	BackEndVN  *appmesh.VirtualNode
+	BackEndDP  *appsv1.Deployment
+	BackEndSVC *corev1.Service
+
+	BackEndVS *appmesh.VirtualService
+	BackEndVR *appmesh.VirtualRouter
+}
+
+func (s *TLSStack) DeployTLSStack(ctx context.Context, f *framework.Framework) {
+	s.createTLSStackMeshAndNamespace(ctx, f)
+	s.ServiceDiscoveryType = manifest.DNSServiceDiscovery
+	mb := &manifest.ManifestBuilder{
+		Namespace:            s.namespace.Name,
+		ServiceDiscoveryType: s.ServiceDiscoveryType,
+	}
+	s.createSecretsForTLSStack(ctx, f, mb)
+	s.createVirtualNodeResourcesForTLSStack(ctx, f, mb)
+	s.createServicesForTLSStack(ctx, f)
+	s.assignBackendVSToFrontEndVN(ctx, f)
+}
+
+func (s *TLSStack) CleanupTLSStack(ctx context.Context, f *framework.Framework) {
+	var deletionErrors []error
+	if errs := s.revokeVirtualNodeBackendAccessForTLSStack(ctx, f); len(errs) != 0 {
+		deletionErrors = append(deletionErrors, errs...)
+	}
+	if errs := s.deleteResourcesForTLSStackServices(ctx, f); len(errs) != 0 {
+		deletionErrors = append(deletionErrors, errs...)
+	}
+	if errs := s.deleteResourcesForTLSStackNodes(ctx, f); len(errs) != 0 {
+		deletionErrors = append(deletionErrors, errs...)
+	}
+	if errs := s.deleteTLSMeshAndNamespace(ctx, f); len(errs) != 0 {
+		deletionErrors = append(deletionErrors, errs...)
+	}
+	for _, err := range deletionErrors {
+		f.Logger.Error("clean up failed", zap.Error(err))
+	}
+	Expect(len(deletionErrors)).To(BeZero())
+}
+
+func (s *TLSStack) DeployPartialTLSStack(ctx context.Context, f *framework.Framework) {
+	s.createTLSStackMeshAndNamespace(ctx, f)
+	time.Sleep(30*time.Second)
```

Comment:
Wondering the need for this sleep. Isn't `createTLSStackMeshAndNamespace` checking for `active` state before returning?